### ISSUE TITLE
question about the detection scores ？

### ISSUE THREAD

**Issue by slz929** [cvlab-epfl/disk#11]
Thanks for your great work!
I wonder why the keypoint scores  are like `[226.70332  214.92577  209.83643  ...  59.53643   59.525322  59.50951 ]`, ?

I noticed that the scores may be the log prob, but log x(0<x<1) will be a negative value, why values above like ..?
And how can I get a prob for the keypoints?

**Comment by jatentaki** [comment]
Hi, that's a good question. Those are unnormalized log probabilities. Recall that in softmax, given unnormalized log probabilities `s`, you compute the actual log-probabilities as `log_p = s - logsumexp(s)`, and then probabilities `p = exp(log_p)` (this is done under the hood of `torch.log_softmax` and `torch.softmax`, respectively). Therefore if you want to obtain log-probabilities of the keypoint being sampled at training time, you'd take the scores for an 8x8 window and subtract their `logsumexp`, which will result in correctly normalized log-probabilities.

At inference time we find that using NMS on those scores is a better way of picking keypoints than probabilitic sampling (mostly because it's not confined to the pre-defined grid) so those scores do not retain their probabilistic interpretation in that case. If you want to obtain meaningful log-probabilities, you can still run `detect.py` with the `--mode rng` flag.

**Comment by slz929** [comment]
> Hi, that's a good question. Those are unnormalized log probabilities. Recall that in softmax, given unnormalized log probabilities `s`, you compute the actual log-probabilities as `log_p = s - logsumexp(s)`, and then probabilities `p = exp(log_p)` (this is done under the hood of `torch.log_softmax` and `torch.softmax`, respectively). Therefore if you want to obtain log-probabilities of the keypoint being sampled at training time, you'd take the scores for an 8x8 window and subtract their `logsumexp`, which will result in correctly normalized log-probabilities.
> 
> At inference time we find that using NMS on those scores is a better way of picking keypoints than probabilitic sampling (mostly because it's not confined to the pre-defined grid) so those scores do not retain their probabilistic interpretation in that case. If you want to obtain meaningful log-probabilities, you can still run `detect.py` with the `--mode rng` flag.

Thanks!
If I just want to obtain the probabilities at inference time, what should i do?

**Comment by jatentaki** [comment]
First, it's a sad fact that in `--mode nms`, which empirically works better on downstream tasks, those scores are not a measure of any actual probability (since nms is not probabilistic). I see two potential approaches you can take:
1. use `--mode rng` if you can accept the suboptimal keypoints as a price for working with actual, well-defined log-probabilities
2. modify the [`nms`](https://github.com/cvlab-epfl/disk/blob/master/disk/model/detector.py#L145) method to include the [tiling](https://github.com/cvlab-epfl/disk/blob/master/disk/model/detector.py#L115) and [normalization](https://github.com/cvlab-epfl/disk/blob/master/disk/model/detector.py#L40) parts of `sample` and replace the unnormalized detection scores currently [returned here](https://github.com/cvlab-epfl/disk/blob/master/disk/model/detector.py#L160) (while still retaining the original nms logic for picking keypoint locations). In other words, compute the selection criteria for both `nms` and `sample`, pick keypoints according to the former and return logp according to the latter. This way you'll get the high quality NMS detections with "counterfactual" scores which would be interpretable as "the log-probability of picking that keypoint under `--mode rng` (even though it was actually selected with nms)".

I suppose option 1. may be better if you're looking to compute something mathy that really depends on those log-probabilities being well defined and option 2. may be better if you want the best feature quality and a bit better defined scores. If you share your use-case I may be able to advise you better.

**Comment by slz929** [comment]
> First, it's a sad fact that in `--mode nms`, which empirically works better on downstream tasks, those scores are not a measure of any actual probability (since nms is not probabilistic). I see two potential approaches you can take:
> 
> 1. use `--mode rng` if you can accept the suboptimal keypoints as a price for working with actual, well-defined log-probabilities
> 2. modify the [`nms`](https://github.com/cvlab-epfl/disk/blob/master/disk/model/detector.py#L145) method to include the [tiling](https://github.com/cvlab-epfl/disk/blob/master/disk/model/detector.py#L115) and [normalization](https://github.com/cvlab-epfl/disk/blob/master/disk/model/detector.py#L40) parts of `sample` and replace the unnormalized detection scores currently [returned here](https://github.com/cvlab-epfl/disk/blob/master/disk/model/detector.py#L160) (while still retaining the original nms logic for picking keypoint locations). In other words, compute the selection criteria for both `nms` and `sample`, pick keypoints according to the former and return logp according to the latter. This way you'll get the high quality NMS detections with "counterfactual" scores which would be interpretable as "the log-probability of picking that keypoint under `--mode rng` (even though it was actually selected with nms)".
> 
> I suppose option 1. may be better if you're looking to compute something mathy that really depends on those log-probabilities being well defined and option 2. may be better if you want the best feature quality and a bit better defined scores. If you share your use-case I may be able to advise you better.

Thanks. As you said, maybe i just want to use scores as probabilities form to as  (a kind of [0,1] normalized) input feature of MLP encoder in Superglue, so maybe option 1 is enough and I will try this option. 
Thank you sincerely, and any results will be posted here.

**Comment by jatentaki** [comment]
In that case, remember that those numbers are the actual logits of the Bernoulli variable which decides whether to keep or reject a proposal (after the categorical-in-grid proposal step, which introduces this whole nms/rng mess). As you can tell they are extremely saturated, but that's technically a valid interpretation. You can therefore decide to forget about the spatial proposal process and just treat them as a measure of confidence that they should be kept.

That said, if you want to learn disk + super glue end to end, you have to use the randomized version (so you get valid log-probabilities for free) and if you want to keep disk frozen, I'd just empirically normalize them with logsoftmax across all features detected in each image or something like that : after all, the super glue will learn how to best interpret them.
